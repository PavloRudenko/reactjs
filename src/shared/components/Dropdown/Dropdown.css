.dropdown {
  position: relative;
}

.list {
  position: absolute;
  z-index: 1;
  top: 100%;
  left: 0;
}
