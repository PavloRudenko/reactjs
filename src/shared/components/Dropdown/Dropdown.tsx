import React, { FC, ReactNode, useEffect, useState } from 'react'

import { concatClasses, noop } from '../../../utils'
import { useIsMounted } from '../../hooks/useIsMounted'

import styles from './Dropdown.css'

interface Props {
  button: ReactNode
  children: ReactNode
  className?: string
  listClassName?: string
  isOpen?: boolean
  onOpen?: () => void
  onClose?: () => void
}

const Dropdown: FC<Props> = ({
  button,
  children,
  className,
  listClassName,
  isOpen,
  onOpen = noop,
  onClose = noop,
}) => {
  const [isDropdownOpen, setIsDropdownOpen] = useState(isOpen)
  useEffect(() => setIsDropdownOpen(isOpen), [isOpen])
  useEffect(() => (isDropdownOpen ? onOpen() : onClose()), [isDropdownOpen])

  const handleOpen = () => {
    if (isOpen === undefined) {
      setIsDropdownOpen(!isDropdownOpen)
    }
  }

  return (
    <div className={concatClasses(styles.dropdown, className)}>
      <div id="test_dropdown" onClick={handleOpen}>
        {button}
      </div>
      {isDropdownOpen && (
        <ul
          className={concatClasses(styles.list, listClassName)}
          onClick={() => setIsDropdownOpen(false)}
        >
          {children}
        </ul>
      )}
    </div>
  )
}

export default Dropdown
