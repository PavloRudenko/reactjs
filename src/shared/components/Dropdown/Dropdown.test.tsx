import React from 'react'
import { shallow, ShallowWrapper } from 'enzyme'

import Dropdown from './Dropdown'

const createElement = () =>
  shallow(
    <Dropdown button={<button>test</button>}>
      <div />
    </Dropdown>
  )

describe('<Dropdown />', () => {
  let wrapper: ShallowWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>

  beforeEach(() => {
    wrapper = createElement()
  })

  test('should render', () => {
    expect(wrapper).toBeDefined()
  })
})
