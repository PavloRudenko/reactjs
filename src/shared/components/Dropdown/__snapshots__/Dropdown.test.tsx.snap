// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown /> create snapshot 1`] = `
<div
  className="dropdown"
>
  <div
    id="test_dropdown"
    onClick={[Function]}
  >
    <button>
      test
    </button>
  </div>
</div>
`;

exports[`<Dropdown /> create snapshot with opened dropdown 1`] = `
<div
  className="dropdown"
>
  <div
    id="test_dropdown"
    onClick={[Function]}
  >
    <button>
      test
    </button>
  </div>
  <ul
    className="list"
    onClick={[Function]}
  >
    <div />
  </ul>
</div>
`;
